ledger/catchpointwriter.go

```diff
@@ -138,93 +160,110 @@ func (cw *catchpointWriter) Abort() error {
 	return os.Remove(cw.filePath)
 }
 
+// WriteStep works for a short period of time (determined by stepCtx) to get
+// some more data (accounts/resources/kvpairs) by using readDatabaseStep, and
+// write that data to the open tar file in cw.tar.  The writing is done in
+// asyncWriter, so that it can proceed concurrently with reading the data from
+// the db. asyncWriter only runs long enough to process the data read during a
+// single call to WriteStep, and WriteStep ensures that asyncWriter has finished
+// writing by waiting for it in a defer block, collecting any errors that may
+// have occurred during writing.  Therefore, WriteStep looks like a simple
+// synchronous function to its callers.
 func (cw *catchpointWriter) WriteStep(stepCtx context.Context) (more bool, err error) {
 	// have we timed-out / canceled by that point ?
 	if more, err = hasContextDeadlineExceeded(stepCtx); more || err != nil {
 		return
 	}
 
-	writerRequest := make(chan catchpointFileBalancesChunkV6, 1)
+	writerRequest := make(chan catchpointFileChunkV6, 1)
 	writerResponse := make(chan error, 2)
-	go cw.asyncWriter(writerRequest, writerResponse, cw.balancesChunkNum, cw.numAccountsProcessed)
+	go cw.asyncWriter(writerRequest, writerResponse, cw.chunkNum)
 	defer func() {
+		// For simplicity, all cleanup is done once, here. The writerRequest is
+		// closed, signaling asyncWriter that it can exit, and then
+		// writerResponse is drained, ensuring any problems from asyncWriter are
+		// noted (and that the writing is done).
 		close(writerRequest)
-		// wait for the writerResponse to close.
+	DRAIN:
 		for {
 			select {
 			case writerError, open := <-writerResponse:
 				if open {
 					err = writerError
 				} else {
-					return
+					break DRAIN
 				}
 			}
 		}
+		if !more {
+			// If we're done, close up the tar file and report on size
+			cw.tar.Close()
+			cw.compressor.Close()
+			cw.file.Close()
+			fileInfo, statErr := os.Stat(cw.filePath)
+			if statErr != nil {
+				err = statErr
+			}
+			cw.writtenBytes = fileInfo.Size()
+
+			// These don't HAVE to be closed, since the "owning" tx will be cmmmitted/rolledback
+			cw.accountsIterator.Close()
+			if cw.kvRows != nil {
+				cw.kvRows.Close()
+				cw.kvRows = nil
+			}
+		}
 	}()
 
 	for {
-		// have we timed-out / canceled by that point ?
+		// have we timed-out or been canceled ?
 		if more, err = hasContextDeadlineExceeded(stepCtx); more || err != nil {
 			return
 		}
 
-		if len(cw.balancesChunk.Balances) == 0 {
+		if cw.chunk.empty() {
 			err = cw.readDatabaseStep(cw.ctx, cw.tx)
 			if err != nil {
 				return
 			}
+			// readDatabaseStep yielded nothing, we're done
+			if cw.chunk.empty() {
+				return false, nil
```

Comment:
@jannotti Sanity checking a possible correctness concern:  Can you review and see if you agree that non-naked returns should return `more`?

* As I understand it, setting `false` here does _not_ change the value of `more` seen in the `defer` function execution.  Here is a toy example:  https://go.dev/play/p/_Ub-8V88cDI.
* It's unclear to me if `more` will have the desired value in these non-naked return statements.  Even if it _is_ the desired value, I think question's suggestion is _less_ error prone.

Reply:
Could you try that playground again, but with `more` as a named return value?

Reply:
@jannotti Sorry + resolving - That was my intent and I misconstructed the example.  With a named return value, the behavior matches what I think we expect.

I apologize for the confusion.  I think no change warranted here.